compiler/src/dotty/tools/io/ClassPath.scala

```diff
@@ -242,8 +242,8 @@ abstract class ClassPath {
       case Some((pkg, rest)) =>
         val rep = packages find (_.name == pkg) flatMap (_ findClass rest)
         rep map {
-          case x: ClassRep  => x
-          case x            => throw new FatalError("Unexpected ClassRep '%s' found searching for name '%s'".format(x, name))
+          case x: AnyClassRep => x
+          case x              => throw new FatalError("Unexpected ClassRep '%s' found searching for name '%s'".format(x, name))
```

Comment:
See comment on the ticket #2156. I think it's a bug in scalac. By guess would be it checks whether the test would redundant (which is reasonable) but gets that logic messed up. For instance, here is a situation where the test should not be omitted:

    class Outer {
      trait Base
      case class Inner() extends Base

      val x: Base = ???
      x match {
        case x: Inner => ???
      }
    }